org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/BuildManager.xtend

```diff
@@ -75,20 +81,41 @@ class BuildManager {
             project2deleted.put(projectManager, deleted)
         }
         val sortedDescriptions = sortByDependencies(project2dirty.keySet + project2deleted.keySet)
-        val result = newArrayList()
         for(ProjectDescription it: sortedDescriptions) {
             val projectManager = workspaceManager.getProjectManager(name)
             val projectDirty = project2dirty.get(it).toList
             val projectDeleted = project2deleted.get(it).toList
-            val partialResult = projectManager.doBuild(projectDirty, projectDeleted, result, cancelIndicator)
+            val partialResult = projectManager.doBuild(projectDirty, projectDeleted, unreportedDeltas, cancelIndicator)
             allDirty.addAll(partialResult.affectedResources.map[uri])
             this.dirtyFiles -= projectDirty
             this.deletedFiles -= projectDeleted
-            result.addAll(partialResult.affectedResources)
+            // prior builds could have been canceled, so their result has not been returned,
+            // but the projectManager already has updated its state
+            mergeWithUnreportedDeltas(partialResult.affectedResources)
         }
+        val result = unreportedDeltas
+       	unreportedDeltas = newArrayList
         return result
     }
     
+    protected def mergeWithUnreportedDeltas(List<IResourceDescription.Delta> newDeltas) {
+    	if (unreportedDeltas.empty) {
+    		unreportedDeltas += newDeltas
+    	} else {
+        	val unreportedByUri = unreportedDeltas.toMap[uri]
+        	newDeltas.forEach [ newDelta |
+        		val unreportedDelta = unreportedByUri.get(newDelta.uri)
+        		if (unreportedDelta === null) {
+        			unreportedDeltas.add(newDelta)
+        		} else {
+					unreportedDeltas -= unreportedDelta
+					unreportedDeltas += new DefaultResourceDescriptionDelta(unreportedDelta.old, newDelta.^new)
+        		}
+        	]
+        	unreportedDeltas.clear
```

Comment:
I do not understand this, sorry. Why do we modify the content of the `unreportedDeltas` if we clear it? What does the `else` branch suppose to do?

Reply:
Good catch. Fixed